lib/internal/fs/promises.js

```diff
@@ -663,19 +666,46 @@ async function writeFile(path, data, options) {
   options = getOptions(options, { encoding: 'utf8', mode: 0o666, flag: 'w' });
   const flag = options.flag || 'w';
 
-  if (!isArrayBufferView(data)) {
-    validateStringAfterArrayBufferView(data, 'data');
-    data = Buffer.from(data, options.encoding || 'utf8');
+  if (isStream(data) || isIterable(data)) {
+    const fd = await open(path, flag, options.mode);
+    if (options.signal?.aborted) {
+      throw lazyDOMException('The operation was aborted', 'AbortError');
+    }
+    try {
+      for await (const buf of data) {
+        await fd.write(buf);
```

Comment:
If signal is aborted, we should:

1. Avoid opening the file.
2. Close it if it has been opened.
3. Stop writing the file and iterating the stream/iterator.

```suggestion
    if (options.signal?.aborted) {
      throw lazyDOMException('The operation was aborted', 'AbortError');
    }
    const fd = await open(path, flag, options.mode);
    try {
      if (options.signal?.aborted) {
        throw lazyDOMException('The operation was aborted', 'AbortError');
      }
      for await (const buf of data) {
        if (options.signal?.aborted) {
          throw lazyDOMException('The operation was aborted', 'AbortError');
        }
        await fd.write(buf);
        if (options.signal?.aborted) {
          throw lazyDOMException('The operation was aborted', 'AbortError');
        }
```